embeddings/task/lightning_task/lightning_task.py

```diff
@@ -0,0 +1,182 @@
+        self._aggregate_and_log_metrics(self.train_metrics)
+
+    def validation_epoch_end(self, outputs: List[Any]) -> None:
+        self._aggregate_and_log_metrics(self.val_metrics, prog_bar=True)
+
+    def test_epoch_end(self, outputs: List[Any]) -> None:
+        self._aggregate_and_log_metrics(self.test_metrics)
+
+    def _aggregate_and_log_metrics(
+        self, metrics: MetricCollection, prog_bar: bool = False
+    ) -> Dict[str, float]:
+        metric_values = metrics.compute()
+        metrics.reset()
+        self.log_dict(metric_values, prog_bar=prog_bar)
+        return metric_values
+
+    def configure_optimizers(self) -> Tuple[List[Optimizer], List[Any]]:
+        """Prepare optimizer and schedule (linear warmup and decay)"""
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": self.hparams.weight_decay,
+            },
+            {
+                "params": [
+                    p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(
+            optimizer_grouped_parameters,
+            lr=self.hparams.learning_rate,
+            eps=self.hparams.adam_epsilon,
+        )
+
+        if self.hparams.use_scheduler:
+            lr_schedulers = self.configure_schedulers(optimizer=optimizer)
+        else:
+            lr_schedulers = []
+
+        return [optimizer], lr_schedulers
+
+    def configure_schedulers(self, optimizer: Optimizer) -> List[Dict[str, Any]]:
+        scheduler = get_linear_schedule_with_warmup(
+            optimizer,
+            num_warmup_steps=self.hparams.warmup_steps,
+            num_training_steps=self.total_steps,
+        )
+        return [{"scheduler": scheduler, "interval": "step", "frequency": 1}]
+
+
+class HuggingFaceLightningTask(LightningTask[AutoModel], abc.ABC):
+    def __init__(
+        self,
+        model_name_or_path: str,
+        downstream_model_type: Type["AutoModel"],
+        unfreeze_transformer_from_layer: Optional[int] = None,
+        metrics: Optional[MetricCollection] = None,
+        config_kwargs: Optional[Dict[str, Any]] = None,
+        task_model_kwargs: Optional[Dict[str, Any]] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(
+            metrics=metrics, **task_model_kwargs if task_model_kwargs else {}, **kwargs
+        )
+        self.save_hyperparameters({"downstream_model_type": downstream_model_type.__name__})
+        self.downstream_model_type = downstream_model_type
+        self.config_kwargs = config_kwargs if config_kwargs else {}
+
+    def setup(self, stage: Optional[str] = None) -> None:
+        if stage in ["fit", None]:
+            self.configure_model()
```

Comment:
what if the stage is 'validate' 'test'? the model will not be initialized.

Reply:
wouldn't this overwrite the trained model with its weights?

Reply:
however, they dont do stages here 
https://pytorch-lightning.readthedocs.io/en/stable/_modules/pytorch_lightning/core/hooks.html#DataHooks.setup